language/changelog/file_format/v2/abilities_and_constraints.md

```diff
@@ -0,0 +1,317 @@
+  * If something can be used across module boundaries and it is expressible at the bytecode level, it must be expressible at the source language level
+  * A source language developer might not always have access to the source code for a compiled Move module (the module might have been written in another language)
+  * As such, the source language must be able to load and interface with compiled Modules in some way
+  * This does not mean that the syntax has to be exactly the same (as the Move compiler could interface with it in some other means) but the level of complexity has to be there. The easiest way to ensure this (while still keeping error messages sane) is to have it be expressible in the syntax
+* A struct with no modifiers will satisfy no constraints
+  * This is intuitive as every constraint _adds_ new behavior.
+    * CAN_COPY lets you copy values (directly on locals or indirectly through references)
+    * CAN_DROP lets you drop values (implicitly with `e;` or explicitly with `let _ = e;`)
+    * GLOBAL_STORAGE_VALUE lets the value exist in global storage
+    * GLOBAL_STORAGE_KEY lets the value be used with move_to and other operations
+  * I believe it would be peculiar if you had to add syntax to remove behavior (having to write out !drop or !copy or something)
+  * I think this will line up well with constraints on generics
+    * `<T: copy>` needs a `copy` struct/type
+* It is the least complicated decision while we are unsure how the new system will be used
+  * Adding `hotpotato` as a keyword to the new system is not clean as it has a weird relational hierarchy (`resource <: hotpotato`)
+  * We aren’t entirely sure what sort of programming patterns will be
+  * If just a handful of combinations are used everywhere, we could look into keywording those combinations and removing the rest
+
+TL;DR Exposing each constraint as a keyword is simple while solving existing needs, and can be changed in the future as we learn more.
+
+# Examples and Use Cases
+
+## Signer
+
+### Kind System
+
+```rust
+script {
+    // the reference is used to prevent it ending up in global storage
+    // the reference makes the VM API weird
+    fun main(s: &signer) { ... }
+}
+```
+
+### Constraint System
+
+```rust
+script {
+    // signer: drop
+    // can’t be copied, can’t be stored
+    // allows for fun main(s: vector<signer>) if we wanted
+    // Most functions would still take &signer, but the VM is simplified
+    fun main(s: signer) { ... }
+}
+```
+
+## Linear Capability
+
+### Kind System
+
+```rust
+address 0x42 {
+module M {
+    resource struct LockCap {}
+    fun lock(s: &signer): LockCap { /* access rules */ ... }
+    // Type system cannot guarantee unlock is called
+    fun unlock(c: LockCap) { ... }
+}
+}
+```
+
+### Constraint System
+
+```rust
+address 0x42 {
+module M {
+    struct LockCap {}
+    fun lock(s: &signer): LockCap { /* access rules */ ... }
+    // Type system guarantees unlock is called!
```

Comment:
how about user define a resource and hold the LockCap?

```rust
address 0x43 {
module M2{  
  resource LockCapHolder{
     cap: LockCap
  }
}
}
```

Reply:
Under this new system, that will not be allowed because `LockCap` is not declared with the `Store` ability.

Reply:
> Under this new system, that will not be allowed because `LockCap` is not declared with the `Store` ability.

Got it. But this will break the current Move module and how to compatible?

Reply:
Any compiled Move modules will still load, as the change is backwards compatible 

Source language files will likely need to be rewritten/updated. I'll eventually update the doc here with how that should be done